### PR TITLE
Allow changing the logged step value in validation_step

### LINKED ISSUE #4102: How to log by epoch for both training and validation on 1.0.0rc4 / 1.0.0rc5 / 1.0.0
#### What is your question?

I have been trying out pytorch-lightning 1.0.0rc5 and wanted to log only on epoch end for both training and validation while having in the x-axis the epoch number. I noticed that `training_epoch_end` now does not allow to return anything. Though I noticed that for training I can achieve what I want by doing:

```python
def training_epoch_end(self, outputs):
    loss = compute_epoch_loss_from_outputs(outputs)
    self.log('step', self.trainer.current_epoch)
    self.log('loss', {'train': loss})
```

It sets the `step` to be the epoch number and used for the x-axis just as I wanted. I have not found in the documentation if this is how it is intended to be logged. I am also a bit confused about the result objects. Nevertheless, this code seems quite simple and logical, so I thought this could be one of the possible intended ways of logging per epoch.

I tried to do the same for validation as follows:

```python
def validation_epoch_end(self, outputs):
    loss = compute_epoch_loss_from_outputs(outputs)
    self.log('step', self.trainer.current_epoch)
    self.log('loss', {'valid': loss})
```

However, in the case of validation the x-axis is the number of batches in validation and an additional `step` graph appears in tensorboard.

Based on this I have some questions. Is this an intended way of logging per epoch? If yes, is the idea that the same behavior is obtained for both training and validation? If this is not the intended way of logging per epoch, where can I read about how this is planned for version 1.0.0?

#### What's your environment?

 - OS: Linux
 - Packaging: pip
 - Version: 1.0.0rc5

### LINKED ISSUE #4203: Viewing validation statistics by epoch (on x-axis) broken in Wandb
## 🐛 Bug

Here's the [boring model](https://colab.research.google.com/drive/1h8KbopeAidzcLc8w7uxiOGftaoztJdMV?usp=sharing).

If you want to view the charts in Weights and Biases with epoch on the X-axis, you get a message that there is "no data availible". Viewing with the step/global step on the X-axis still works. See the two images below:

*With epoch on x-axis*
Screenshot from 2020-10-17 15-46-05

*With global_step on x-axis*
Screenshot from 2020-10-17 15-45-37

I suspect this is related to the values that PL sends with the logs, for example this is what the CSVLogger produces:
Screenshot from 2020-10-17 15-47-59

For some strange reason, there is no epoch value for validation data - the place that should definitely have it...

This bug is extremely annoying if you are, for example, experimenting with different batch sizes - then just comparing steps is meaningless...

### Environment

```
* CUDA:
	- GPU:
		- Tesla T4
	- available:         True
	- version:           10.1
* Packages:
	- numpy:             1.18.5
	- pyTorch_debug:     False
	- pyTorch_version:   1.6.0+cu101
	- pytorch-lightning: 1.0.2
	- tqdm:              4.41.1
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         x86_64
	- python:            3.6.9
	- version:           #1 SMP Thu Jul 23 08:00:38 PDT 2020
```

P.S.: Default PL in boring model should be changed to 1.0.2.

### PR DESCRIPTION
**Pull request by mauvilsa**
##  What does this PR do?

This fixes the bug identified in the discussion of https://github.com/PyTorchLightning/pytorch-lightning/issues/4102

Fixes #4102
Fixes #4203

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    

@rohitgr7 please review.

### CODE REVIEW

**Comment by mauvilsa** [comment]
I see that unit tests are now failing. There are asserts that check what has been logged and it now differs because `epoch` is no automatically logged. But I understood that having `epoch` is a feature. What should be done here? Change the tests?

**Comment by rohitgr7** [comment]
> But I understood that having epoch is a feature. What should be done here? Change the tests?

not sure about this. let's see what other reviewers say.

**Comment by williamFalcon** [comment]
i don't think this is a bug... again, you shouldn't really log validation each step... it means nothing.

**Comment by mauvilsa** [comment]
@williamFalcon let me explain the change. This is not about logging validation each step. The hard coding of the `step` argument to `log_metrics` in [logger_connector.py#L202](https://github.com/PyTorchLightning/pytorch-lightning/blob/1.0.0/pytorch_lightning/trainer/connectors/logger_connector.py#L202) is unnecessary since if omitted it would get the same value anyway because of [logger_connector.py#L90](https://github.com/PyTorchLightning/pytorch-lightning/blob/1.0.0/pytorch_lightning/trainer/connectors/logger_connector.py#L90). However, this hard coding prevents the user to override the value of step, making inconsistent the behavior between training and validation.

The only issue is the automatic logging of epoch in [logger_connector.py#L89](https://github.com/PyTorchLightning/pytorch-lightning/blob/1.0.0/pytorch_lightning/trainer/connectors/logger_connector.py#L89) which makes some unit tests fail. Maybe this automatic logging should be off by default and only enabled if requested by the user?

**Comment by rohitgr7** [comment]
I thing #4132 solved this issue, not sure why it was closed.

File: pytorch_lightning/trainer/connectors/logger_connector.py [hunk 1]
```diff
@@ -87,7 +87,7 @@ def log_metrics(self, metrics, grad_norm_dic, step=None):
         elif step is None:
             # added metrics by Lightning for convenience
             scalar_metrics['epoch'] = self.trainer.current_epoch
-            step = step if step is not None else self.trainer.global_step
+            step = self.trainer.global_step
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/connectors/logger_connector.py:90]: I believe #4132 tried to solve the problem of logging epoch on the x-axis in case of `*_epoch_end`, not sure why it was closed. I suggest by default it should set `global_step` for step_logs and `current_epoch` for epoch_logs and in case if `step` is provided in the metrics by the user, it should be replaced with that value.

Comment (mauvilsa) [reply pytorch_lightning/trainer/connectors/logger_connector.py:90]: I agree. Even better if for epoch_logs the default is the current epoch.

Comment (mauvilsa) [reply pytorch_lightning/trainer/connectors/logger_connector.py:90]: From feedback it is clear that the default should be `global_step`.

Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/logger_connector.py:90]: please note that the log_metrics method here is recursive. 
At the bottom, it is calling itself with step (before) but now step is None. This means in the recursive call, step will be None and then it will add the epoch key. That is why you see the epoch key in the test as extra. Just wanted to point that out in case this was not clear.

**Comment by awaelchli** [comment]
I think the automatic epoch logging should stay. It is very useful to have that by default, because it tells us how many steps correspond to one epoch, and in advanced loggers it allows us to change the axis of plots to epoch. So my suggestion is to adjust the tests accordingly.

File: tests/trainer/logging/test_eval_loop_logging_1_0.py [hunk 2]
```diff
@@ -198,10 +199,11 @@ def validation_epoch_end(self, outputs):
     assert pbar_metrics == expected_pbar_metrics
 
     callback_metrics = set(trainer.callback_metrics.keys())
+    callback_metrics.remove('debug_epoch')
```
Comment (awaelchli) [inline tests/trainer/logging/test_eval_loop_logging_1_0.py:202]: anyone knows what this is and what it is needed for?

**Review by awaelchli — APPROVED** [review]
works! the epoch is back <3

File: tests/trainer/logging/test_eval_loop_logging_1_0.py (hunk 2 above)
Comment (rohitgr7) [reply tests/trainer/logging/test_eval_loop_logging_1_0.py:202]: Internal debugger https://github.com/PyTorchLightning/pytorch-lightning/blob/cf9fe4905eac1bb66a8b72c7799cf77d6f6450f3/pytorch_lightning/trainer/connectors/logger_connector.py#L102-L109

**Review by justusschock — APPROVED** [review]

**Review by rohitgr7 — APPROVED** [review]
@mauvilsa thanks for the fix. Also next time pls make a branch(not master) and push? would be easier to test locally :)
